BUG: guard against replacing constants without `'_'` spec

utke1 opened the pull request:
In the reported problem snippet the attempt to infer the type of
    parameter (wasize=maxiterates*2)
leads to a lookup of maxiterates in the logic of get_parameters which finds
and correctly assigns to 'v' the value string '50000'
However, then in the logic proceds to attempt to split off any underscore-appended precision
spec from v_ which yields ['50000']  because there is no underscore.
Finally line
v = ''.join(v_[:-1]).lower().replace(v_[-1].lower(), '')
if not quarded by the newly introduced lengthtest sets v to empty because
[:-1]  of a list with one element yields an empty list.
Subsequently the type inference by running eval on an empty string fails with the error
message quoted.
The introduced length check appears to correct this problem.

Commit:
BUG: guard against replacing constants without '_' spec

fixes #8493
In the reported problem snippet the attempt to infer the type of
    parameter (wasize=maxiterates*2)
leads to a lookup of maxiterates in the logic of get_parameters
which finds and correctly assigns to 'v' the value string '50000'
However, then in the logic proceds to attempt to split off any
underscore-appended precision spec from v_ which yields ['50000']
because there is no underscore.
Finally line
v = ''.join(v_[:-1]).lower().replace(v_[-1].lower(), '')
if not quarded by the newly introduced lengthtest sets v to empty
because [:-1]  of a list with one element yields an empty list.
Subsequently the type inference by running eval on an empty string
fails with the error message quoted.
The introduced length check appears to correct this problem.

charris commented:
Needs tests, should also test if this fixes #8479. Also read `doc/source/dev/gitwash/development_workflow.rst` on formatting commit messages. Note that they need hard line breaks and the lines should be <= 72 characters long.

charris commented:
And the `Fixes #8493` should be part of the commit message rather than the summary line.

charris commented:
You can see some tests at #8299. @zerothi Ping.

utke1 commented:
To @charris  - when you commented on the formatting of the commit message - do you require me to re-commit the change (i.e. kill this pull request and make a new one) or will you accept the commit message as is and I will just add some test cases into this pull request?

charris commented:
You can edit commit messages, squash commits, and do other good things using `git commit -i HEAD~n`, where n is the number of recent commits to edit, followed by a force push `git push -f origin HEAD`.

pv commented:
@charris probably meant `git rebase -i HEAD~n` not `git commit ...`.

madphysicist commented:
Also, keep in mind that the PR is tied to the branch name, so any sequence of commits you assign the name to will be the subject of this PR. Modify your git history at will with rebasing, cherry-picking or outright branch renaming with that in mind.

charris commented:
Ah, I see that you are also working in the master branch, big no no. You can fix that by checking out a work branch `git co -b <new branch>`, then resetting master to before your work `git reset --hard <hash of last valid commit>. You will need to close this PR and make a new one.

zerothi commented:
I have went through the code and it looks good to me, thanks for fixing my mistake. I have added a test which you may grab here:
<script src="https://gist.github.com/zerothi/d7f7beb5064be69281a3b0ede97a769d.js"></script>

utke1 commented:
Changed commit message and pull request title to match top line commit message

Commit:
TST: adding tests for compound constant provided by @zerothi

This is a test code provided as a patch by @zerothi checking
the compound constant parsing.

Commit:
TST: adding test for constants without compound kind spec

This augments the test in constant_compound.f90 by
using constants without a compound kind spec to
illustrate the case that led to the
reporting of issue #8493

utke1 commented:
To @charris  - with the 2 tests included above - this is all I was going to do to amend this pull request - please let me know if there is anything else you need me to do before you can reconsider merging.

charris commented on numpy/f2py/crackfortran.py:
```diff
@@ -2451,7 +2451,8 @@ def get_parameters(vars, global_params={}):
                     v_ = v.split('_')
                     # In case there are additive parameters
-                    v = ''.join(v_[:-1]).lower().replace(v_[-1].lower(), '')
+                    if len(v_)>1: 
```
Spaces around `>`.

utke1 replied:
done

charris commented:
Just a small style nitpick. @2sn Does this fix your problem?

Commit:
STY: space around operator

per request from @charris

charris commented:
Doesn't seem to be anymore feedback on this, so in it goes. Thanks @utke1 .

charris commented:
Might want to clean up your master branch now as it will be out of sync with numpy master. Or you could just refork and start over.

2sn commented:
Yes, this seems to fix the compile issue. Thanks!
